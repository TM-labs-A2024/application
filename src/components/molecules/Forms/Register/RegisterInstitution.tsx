/* eslint-disable no-console */
import { Button, FormControl, FormErrorMessage, Input, Heading, Stack } from '@chakra-ui/react'
import Splash from '@components/atoms/Splash'
import { institutionTypes } from '@components/molecules/Forms/Register/Register.constants'
import { InstitutionRegister, Login, ReactSelectOption } from '@src/types'
import React, { ReactElement, useState, useEffect } from 'react'
import { Controller, useForm, FieldErrors } from 'react-hook-form'
import Select from 'react-select'

import Logo from '../../../../../public/static/icons/logo.svg'

const institutionTypeOptions = institutionTypes.map((option) => ({
  value: option.name,
  label: option.name
}))

type FormData = {
  address: string
  firstname: string
  lastname: string
  govId: string
  birthdate: string
  credentials: string
  email: string
  phoneNumber: string
  password: string
  repeatPassword: string
  code: string
  title: string
  institution: string
  institutionId: string
  institutionType: ReactSelectOption | null
}

function RegisterHeader() {
  return (
    <>
      <Logo className="mx-auto my-4 hidden md:block" />
      <Heading as="h2" size="md" mb={4} noOfLines={1}>
        Registro
      </Heading>
    </>
  )
}

export default function RegisterForm({
  context: { createInstitution, loginInstitution, verificationCode, institutionCreated, isLoading }
}: {
  context: {
    createInstitution: (arg: InstitutionRegister) => void
    loginInstitution: (arg: Login) => void
    verificationCode: string
    institutionCreated: boolean
    isLoading: boolean
  }
}): ReactElement {
  // --- Hooks -----------------------------------------------------------------
  const {
    handleSubmit,
    formState: { errors, isSubmitting },
    control,
    register
  } = useForm<FormData>()
  // --- END: Hooks ------------------------------------------------------------

  // --- Local state -----------------------------------------------------------
  const [step, setStep] = useState(1)
  // --- END: Local state ------------------------------------------------------

  // --- Refs ------------------------------------------------------------------
  // --- END: Refs -------------------------------------------------------------

  // --- Redux -----------------------------------------------------------------
  // --- END: Redux ------------------------------------------------------------

  // --- Side effects ----------------------------------------------------------
  // --- END: Side effects -----------------------------------------------------

  // --- Data and handlers -----------------------------------------------------
  const onSubmitDetails = (data: FormData) => {
    console.log(JSON.stringify(data))
    setStep(2)
  }

  const onSubmitUserDetails = (data: FormData) => {
    console.log(JSON.stringify(data))
    setStep(3)
  }

  const onSubmitPassword = (data: FormData) => {
    const body = {
      name: data.institution,
      govId: data.institutionId,
      credentials: data.credentials,
      type: String(data?.institutionType?.value).toLocaleLowerCase(),
      address: data.address,
      institutionUser: {
        firstname: data.firstname,
        lastname: data.lastname,
        govId: data.govId,
        birthdate: data.birthdate,
        email: data.email,
        password: data.password,
        phoneNumber: data.phoneNumber,
        role: data.title
      }
    }

    createInstitution(body)
  }

  const onSubmitConfirmation = (data: FormData) => {
    loginInstitution(data)
  }

  const verifyErrors = (errors: FieldErrors<FormData>) => Object.keys(errors).length > 0
  // --- END: Data and handlers ------------------------------------------------

  // --- Side effects ----------------------------------------------------------
  useEffect(() => {
    if (institutionCreated) {
      setStep(4)
    }
  }, [institutionCreated])
  // --- END: Side effects -----------------------------------------------------

  return isLoading ? (
    <Splash />
  ) : (
    <div className="h-full w-full px-8 py-12 lg:px-96">
      {step === 1 && (
        <form
          onSubmit={handleSubmit(onSubmitDetails)}
          className="h-full w-full"
          data-testid="register-institution-step-1"
        >
          <FormControl
            isInvalid={verifyErrors(errors)}
            className="flex h-full w-full flex-col justify-between"
          >
            <RegisterHeader />
            <Heading as="h2" size="sm" mb={4} noOfLines={1}>
              Datos de la institución
            </Heading>
            <Stack className="h-3/4 overflow-scroll">
              <Input
                id="institution"
                className="mt-2 min-h-10"
                placeholder="Nombre de la institución"
                {...register('institution', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>
                {errors?.institution && errors?.institution?.message}
              </FormErrorMessage>
              <Input
                id="institutionId"
                className="mt-2 min-h-10"
                placeholder="RIF"
                {...register('institutionId', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>
                {errors?.institutionId && errors?.institutionId?.message}
              </FormErrorMessage>
              <Input
                id="address"
                className="mt-2 min-h-10"
                placeholder="Dirección"
                {...register('address', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>{errors?.address && errors?.address?.message}</FormErrorMessage>
              <Input
                id="credentials"
                className="mt-2 min-h-10"
                placeholder="Credenciales"
                {...register('credentials', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>
                {errors?.credentials && errors?.credentials?.message}
              </FormErrorMessage>
              <Controller
                control={control}
                name="institutionType"
                rules={{
                  required: 'Este campo es obligatorio'
                }}
                render={({ field }) => (
                  <Select
                    id="institutionType"
                    {...field}
                    placeholder="Tipo de institución"
                    options={institutionTypeOptions}
                    className="mt-2"
                  />
                )}
              />
              <FormErrorMessage>
                {errors?.institutionType && errors?.institutionType?.message}
              </FormErrorMessage>
            </Stack>
            <Button isLoading={isSubmitting} type="submit">
              Siguiente
            </Button>
          </FormControl>
        </form>
      )}
      {step === 2 && (
        <form onSubmit={handleSubmit(onSubmitUserDetails)} className="h-full w-full">
          <FormControl
            isInvalid={verifyErrors(errors)}
            className="flex h-full w-full flex-col justify-between"
          >
            <RegisterHeader />
            <Heading as="h2" size="sm" mb={4} noOfLines={1}>
              Datos de administración
            </Heading>
            <Stack className="h-3/4 overflow-scroll">
              <Input
                id="firstname"
                className="mt-2 min-h-10"
                placeholder="Nombre"
                {...register('firstname', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>{errors?.firstname && errors?.firstname?.message}</FormErrorMessage>
              <Input
                id="lastname"
                className="mt-2 min-h-10"
                placeholder="Apellido"
                {...register('lastname', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>{errors?.lastname && errors?.lastname?.message}</FormErrorMessage>
              <Input
                id="govId"
                className="mt-2 min-h-10"
                placeholder="Cédula"
                {...register('govId', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>{errors?.govId && errors?.govId?.message}</FormErrorMessage>
              <Input
                id="birthdate"
                className="mt-2 min-h-10"
                type="date"
                placeholder="Fecha de nacimiento"
                {...register('birthdate', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>{errors?.birthdate && errors?.birthdate?.message}</FormErrorMessage>
              <Input
                id="title"
                className="mt-2 min-h-10"
                placeholder="Cargo"
                {...register('title', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>{errors?.title && errors?.title?.message}</FormErrorMessage>
              <Input
                id="phoneNumber"
                className="mt-2 min-h-10"
                placeholder="Teléfono"
                {...register('phoneNumber', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>
                {errors?.phoneNumber && errors?.phoneNumber?.message}
              </FormErrorMessage>
              <Input
                id="email"
                className="mt-2 min-h-10"
                placeholder="Correo electrónico"
                {...register('email', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>{errors?.email && errors?.email?.message}</FormErrorMessage>
            </Stack>
            <Button isLoading={isSubmitting} type="submit">
              Siguiente
            </Button>
          </FormControl>
        </form>
      )}
      {step === 3 && (
        <form
          onSubmit={handleSubmit(onSubmitPassword)}
          className="h-full w-full"
          data-testid="register-step-2"
        >
          <FormControl
            isInvalid={verifyErrors(errors)}
            className="flex h-full w-full flex-col justify-between"
          >
            <RegisterHeader />
            <Heading as="h2" size="sm" mb={4} noOfLines={1}>
              Seguridad
            </Heading>
            <Stack className="h-3/4 overflow-scroll">
              <Input
                id="password"
                className="mt-2 min-h-10"
                type="password"
                placeholder="Contraseña"
                {...register('password', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>{errors?.password && errors?.password?.message}</FormErrorMessage>
              <Input
                id="repeatPassword"
                className="mt-2 min-h-10"
                type="password"
                placeholder="Repetir contraseña"
                {...register('repeatPassword', {
                  required: 'Este campo es obligatorio'
                })}
              />
              <FormErrorMessage>
                {errors?.repeatPassword && errors?.repeatPassword?.message}
              </FormErrorMessage>
            </Stack>
            <Button isLoading={isSubmitting} type="submit">
              Siguiente
            </Button>
          </FormControl>
        </form>
      )}
      {step === 4 && (
        <form onSubmit={handleSubmit(onSubmitConfirmation)} className="h-full w-full">
          <FormControl
            isInvalid={verifyErrors(errors)}
            className="flex h-full w-full flex-col justify-between"
          >
            <RegisterHeader />
            <Heading as="h2" size="sm" mb={4} noOfLines={1}>
              Confirmar correo electronico
            </Heading>
            <Stack className="h-3/4 overflow-scroll">
              <Input
                id="code"
                className="mt-2 min-h-10"
                placeholder="Ingresar código"
                {...register('code', {
                  required: 'Este campo es obligatorio',
                  validate: {
                    verifyCode: (code) => verificationCode === code || 'El código no concuerda'
                  }
                })}
              />
              <FormErrorMessage>{errors?.code && errors?.code?.message}</FormErrorMessage>
            </Stack>
            <Button isLoading={isSubmitting} type="submit">
              Registrar institución
            </Button>
          </FormControl>
        </form>
      )}
    </div>
  )
}
