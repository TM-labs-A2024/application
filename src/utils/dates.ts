export const formatDate = (date: string) => {
  const chunks = date.split('-')

  const newDate = chunks[1] + '-' + chunks[2] + '-' + chunks[0]

  return newDate
}
