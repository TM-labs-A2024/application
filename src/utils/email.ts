import emailjs from '@emailjs/browser'

export function sendEmail(data: {
  from_name: string
  to_name: string
  code: string | number
  to_email: string
}) {
  const templateParams = {
    ...data
  }

  const serviceKey = process.env.NEXT_PUBLIC_EMAIL_SERVICE_KEY ?? ''
  const templateKey = process.env.NEXT_PUBLIC_EMAIL_TEMPLATE_KEY ?? ''
  const apiKey = process.env.NEXT_PUBLIC_EMAIL_API_KEY ?? ''

  emailjs.send(serviceKey, templateKey, templateParams, apiKey).catch(() => {
    alert('Algo sali√≥ mal, por favor reintenta el registro.')
  })
}
