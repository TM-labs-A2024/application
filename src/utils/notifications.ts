import { GENERIC_ERROR } from '@constants/index'
import { isMobile } from '@utils/index'
import { Store } from 'react-notifications-component'

export const setupErrorNotification = () => {
  Store.addNotification(
    GENERIC_ERROR('Algo salió mal, por favor inténtelo de nuevo.', isMobile(window))
  )
}
