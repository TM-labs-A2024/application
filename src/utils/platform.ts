import { Capacitor } from '@capacitor/core'

export const isWeb = () => Capacitor.getPlatform() === 'web'

export const isIOS = () => Capacitor.getPlatform() === 'ios'

export const isAndroid = () => Capacitor.getPlatform() === 'android'
