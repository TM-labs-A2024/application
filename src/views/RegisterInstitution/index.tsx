import { RegisterInstitutionForm } from '@components/molecules/Forms/Register'
import { InstitutionRegister, Login } from '@src/types'
import React from 'react'

export default function RegisterInstitutionView({
  context
}: {
  context: {
    createInstitution: (arg: InstitutionRegister) => void
    loginInstitution: (arg: Login) => void
    verificationCode: string
    institutionCreated: boolean
    isLoading: boolean
  }
}) {
  return (
    <div className="flex h-screen w-screen flex-col items-center justify-between">
      <RegisterInstitutionForm context={context} />
    </div>
  )
}
