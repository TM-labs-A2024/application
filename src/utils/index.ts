import { formatDate } from './dates'
import { sendEmail, generateVerificationCode, setupEmailSending } from './email'
import { setupErrorNotification } from './notifications'
import { isWeb, isIOS, isAndroid, isMobile } from './platform'

export {
  isWeb,
  isIOS,
  isAndroid,
  isMobile,
  sendEmail,
  generateVerificationCode,
  setupEmailSending,
  setupErrorNotification,
  formatDate
}
