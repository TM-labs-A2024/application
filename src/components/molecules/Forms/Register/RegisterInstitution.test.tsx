import { render, screen } from '@test/utils/index'
import 'next/navigation'

import RegisterInstitution from './RegisterInstitution'
import '@testing-library/jest-dom'

jest.mock('next/navigation')

describe('Molecules > Forms > RegisterInstitution test', () => {
  test('The component requires the fields for a patient in step 1', () => {
    render(<RegisterInstitution />)

    expect(screen.getByPlaceholderText('Nombre de la institución')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('RIF')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('Credenciales')).toBeInTheDocument()
    expect(screen.getByText('Tipo de institución')).toBeInTheDocument()
  })

  test('Step 1 matches the snapshot', () => {
    render(<RegisterInstitution />)

    expect(screen.getByTestId('register-institution-step-1')).toMatchSnapshot(
      'register-institution-step-1'
    )
  })
})
