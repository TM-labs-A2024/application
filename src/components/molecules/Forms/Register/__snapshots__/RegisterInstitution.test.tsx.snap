// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[
  `Molecules > Forms > RegisterInstitution test Step 1 matches the snapshot: register-institution-step-1 1`
] = `
<form
  class="h-full w-full"
  data-testid="register-institution-step-1"
>
  <div
    class="chakra-form-control flex h-full w-full flex-col justify-between css-1kxonj9"
    role="group"
  >
    <img
      alt="logo"
      class="mx-auto my-4 hidden md:block"
      data-nimg="1"
      decoding="async"
      height="80"
      loading="lazy"
      src="/_next/image?url=%2Fstatic%2Fimages%2Flogo-horizontal.png&w=640&q=75"
      srcset="/_next/image?url=%2Fstatic%2Fimages%2Flogo-horizontal.png&w=256&q=75 1x, /_next/image?url=%2Fstatic%2Fimages%2Flogo-horizontal.png&w=640&q=75 2x"
      style="color: transparent;"
      width="200"
    />
    <h2
      class="chakra-heading css-170f0s2"
    >
      Registro
    </h2>
    <h2
      class="chakra-heading css-89pyn4"
    >
      Datos de la institución
    </h2>
    <div
      class="chakra-stack h-3/4 overflow-scroll css-1811skr"
    >
      <input
        class="chakra-input min-h-10 css-1cjy4zv"
        id="institution"
        name="institution"
        placeholder="Nombre de la institución"
      />
      <input
        class="chakra-input min-h-10 css-1cjy4zv"
        id="institutionId"
        name="institutionId"
        placeholder="RIF"
      />
      <input
        class="chakra-input min-h-10 css-1cjy4zv"
        id="credential"
        name="credential"
        placeholder="Credenciales"
      />
      <div
        class=" css-b62m3t-container"
        id="institutionType"
      >
        <span
          class="css-1f43avz-a11yText-A11yText"
          id="react-select-3-live-region"
        />
        <span
          aria-atomic="false"
          aria-live="polite"
          aria-relevant="additions text"
          class="css-1f43avz-a11yText-A11yText"
          role="log"
        />
        <div
          class=" css-13cymwt-control"
        >
          <div
            class=" css-1fdsijx-ValueContainer"
          >
            <div
              class=" css-1jqq78o-placeholder"
              id="react-select-3-placeholder"
            >
              Tipo de institución
            </div>
            <div
              class=" css-qbdosj-Input"
              data-value=""
            >
              <input
                aria-activedescendant=""
                aria-autocomplete="list"
                aria-describedby="react-select-3-placeholder"
                aria-expanded="false"
                aria-haspopup="true"
                autocapitalize="none"
                autocomplete="off"
                autocorrect="off"
                class=""
                id="react-select-3-input"
                role="combobox"
                spellcheck="false"
                style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                tabindex="0"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          >
            <span
              class=" css-1u9des2-indicatorSeparator"
            />
            <div
              aria-hidden="true"
              class=" css-1xc3v61-indicatorContainer"
            >
              <svg
                aria-hidden="true"
                class="css-tj5bde-Svg"
                focusable="false"
                height="20"
                viewBox="0 0 20 20"
                width="20"
              >
                <path
                  d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                />
              </svg>
            </div>
          </div>
        </div>
        <input
          name="institutionType"
          type="hidden"
          value=""
        />
      </div>
    </div>
    <button
      class="chakra-button css-ez23ye"
      type="submit"
    >
      Siguiente
    </button>
  </div>
</form>
`
